docs/ARCHITECTURE.md

```diff
@@ -0,0 +1,76 @@
+# Architecture
+
+This library is under development, and it's primarily designed to fulfil needs of the
+open telemetry exporter and therefore some functionality that would be useful in other
+scenarios is not implemented.
+
+## Idea
+
+Design is heavily influenced by peculiarities of the upstream and downstream systems
+about which I know very little.
+
+### Downstream
+
+The downstream system is DataSet REST API and its API call [addEvents](https://app.scalyr.com/help/api#addEvents).
+It has the following properties:
+
+* `D1` - only one request for each session could be made
+* `D2` - when new version is deployed there is few minutes outage
+* `D3` - maximum payload size for the API is 6MB before compression
+
+### Upstream
+
+The upstream system is [open telemetry collector](https://opentelemetry.io/docs/collector/),
+where the datasetexporter is one of many exporters that could be used
+(on the right side of the image). It has the following properties:
+
+* `U1` - data are passed in batches
+* `U2` - each batch could be either accepted or rejected, if it's accepted it cannot be returned, if it's rejected, it may be retried
+* `U3` - function for accepting batches could be called in parallel (with different batches)
+
+### Consequences
+
+As a consequence of the upstream and downstream system following has to happen (codes
+between brackets points to the previous points):
+
+* `C1` - data has to be grouped to be divided into multiple sessions to increase throughput (`D1`)
+* `C2` - data from multiple input batches has to be combined (`D1`, `U1`)
+* `C3` - input and output has to be decoupled (`D1`, `C1`, `C2`)
+* `C4` - if the downstream returns error, system cannot return data that has already accepted, it can only start rejecting new data (`U2`, `U3`)
+* `C5` - since the system cannot return data, it has to handle retries by itself (`C4`)
+* `C6` - user has to be able to define the grouping mechanism to achieve maximum throughput (`C1`)
+
+## Implementation
+
+* `I1` - to decouple upstream from downstream we use pub/sub pattern (`C3`)
+* `I2` - there are two levels of pub/sub patterns
+  * `I2.1` - publisher 1 - events from incoming batch are divided based on the grouping into sessions and published into the corresponding topic (`I1`, `C6`)
```

Comment:
Based on the code its not session but keys. Session = ClientId + Key

Reply:
This is really great question!!! 

Since the `group_by` may not be specified or it's mentioning same key referencing the same file - `/var/log/apache/access. log` - it would lead to multiple instances of the client using the same key => breaking the requirement of having single session in progress. Therefore each client has unique key - and then is then combined with the computed key based on `group_by`.